embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/worker/PriorityRunnable.kt

```diff
@@ -0,0 +1,55 @@
+package io.embrace.android.embracesdk.worker
+
+import java.util.concurrent.Callable
+import java.util.concurrent.RunnableFuture
+
+/**
+ * An implementation of [RunnableFuture] that also contains priority information on how important
+ * the task is.
+ */
+internal class PriorityRunnableFuture<T> (
+    val impl: RunnableFuture<T>,
+    val priority: TaskPriority,
+    val submitTime: Long
+) : RunnableFuture<T> by impl
+
+/**
+ * An implementation of [Runnable] that also contains priority information on how important the
+ * task is.
+ */
+internal class PriorityRunnable(
+    val priority: TaskPriority,
+    impl: Runnable
+) : Runnable by impl
+
+/**
+ * An implementation of [Callable] that also contains priority information on how important the
+ * task is.
+ */
+internal class PriorityCallable<T>(
+    val priority: TaskPriority,
+    impl: Callable<T>
+) : Callable<T> by impl
+
+/**
+ * The priority of a task submitted to the [BackgroundWorker].
+ *
+ * Tasks with higher priority will be executed first.
+ */
+internal enum class TaskPriority(
+
+    /**
+     * The delay threshold in milliseconds that should be added to the task's submit time
+     * when comparing queue priority.
+     *
+     * The higher the priority, the lower the delay threshold.
+     *
+     * See [PriorityThreadPoolExecutor] for further detail.
+     */
+    val delayThresholdMs: Long
+) {
+    CRITICAL(0),
+    HIGH(5_000),
+    NORMAL(15_000),
```

Comment:
Maybe something to think about - 15s isn't a tonne of delay when we are talking about network requests running serially, especially in network conditions that will cause a backlog of queued requests. Bumping this to a higher number might make this work better in practice.

Reply:
What sort of delay do you think would be appropriate? Perhaps `30` for NORMAL and `60` for LOW?